sympy/parsing/tests/test_mathematica.py

```diff
@@ -64,3 +71,156 @@ def test_mathematica():
         assert mathematica(e) == sympify(d[e])
+
+
+def test_parser_mathematica_tokenizer():
+    parser = MathematicaParser()
+
+    chain = lambda expr: parser._parse_tokenized_code(parser._tokenize_mathematica_code(expr))
+
+    # Basic patterns
+    assert chain("x") == "x"
+    assert chain("42") == "42"
+    assert chain(".2") == ".2"
+    assert chain("+x") == "x"
+    assert chain("-1") == "-1"
+    assert chain("- 3") == "-3"
+    assert chain("+Sin[x]") == ["Sin", "x"]
+    assert chain("-Sin[x]") == ["Times", "-1", ["Sin", "x"]]
+    assert chain("x(a+1)") == ["Times", "x", ["Plus", "a", "1"]]
+    assert chain("(x)") == "x"
+    assert chain("(+x)") == "x"
+    assert chain("-a") == ["Times", "-1", "a"]
+    assert chain("(-x)") == ["Times", "-1", "x"]
+    assert chain("(x + y)") == ["Plus", "x", "y"]
+    assert chain("3 + 4") == ["Plus", "3", "4"]
+    assert chain("a - 3") == ["Plus", "a", "-3"]
+    assert chain("a - b") == ["Plus", "a", ["Times", "-1", "b"]]
+    assert chain("7 * 8") == ["Times", "7", "8"]
+    assert chain("a + b*c") == ["Plus", "a", ["Times", "b", "c"]]
+    assert chain("a + b* c* d + 2 * e") == ["Plus", "a", ["Times", "b", "c", "d"], ["Times", "2", "e"]]
+    assert chain("a / b") == ["Times", "a", ["Power", "b", "-1"]]
+
+    # Missing asterisk (*) patterns:
+    assert chain("x y") == ["Times", "x", "y"]
+    assert chain("3 4") == ["Times", "3", "4"]
+    assert chain("a[b] c") == ["Times", ["a", "b"], "c"]
+    assert chain("(x) (y)") == ["Times", "x", "y"]
+    assert chain("3 (a)") == ["Times", "3", "a"]
+    assert chain("(a) b") == ["Times", "a", "b"]
+    assert chain("4.2") == "4.2"
+    assert chain("4 2") == ["Times", "4", "2"]
+    assert chain("4  2") == ["Times", "4", "2"]
+    assert chain("3 . 4") == ["Dot", "3", "4"]
+    assert chain("4. 2") == ["Times", "4.", "2"]
+    assert chain("x.y") == ["Dot", "x", "y"]
+    assert chain("4.y") == ["Times", "4.", "y"]
+    assert chain("4 .y") == ["Dot", "4", "y"]
+    assert chain("x.4") == ["Times", "x", ".4"]
+    assert chain("x0.3") == ["Times", "x0", ".3"]
+    assert chain("x. 4") == ["Dot", "x", "4"]
+
+    # Operators couples + and -, * and / are mutually associative:
+    # (i.e. expression gets flattened when mixing these operators)
+    assert chain("a*b/c") == ["Times", "a", "b", ["Power", "c", "-1"]]
+    assert chain("a/b*c") == ["Times", "a", ["Power", "b", "-1"], "c"]
+    assert chain("a+b-c") == ["Plus", "a", "b", ["Times", "-1", "c"]]
+    assert chain("a-b+c") == ["Plus", "a", ["Times", "-1", "b"], "c"]
+    assert chain("-a + b -c ") == ["Plus", ["Times", "-1", "a"], "b", ["Times", "-1", "c"]]
+    assert chain("a/b/c*d") == ["Times", "a", ["Power", "b", "-1"], ["Power", "c", "-1"], "d"]
+    assert chain("a/b/c") == ["Times", "a", ["Power", "b", "-1"], ["Power", "c", "-1"]]
+    assert chain("a-b-c") == ["Plus", "a", ["Times", "-1", "b"], ["Times", "-1", "c"]]
+    assert chain("1/a") == ["Times", "1", ["Power", "a", "-1"]]
```

Comment:
@certik @oscarbenjamin do you have access to Wolfram Mathematica?

Can you please check if the output of `FullForm[Hold[1/a]]` is `Times[1, Power[a, -1]]` or `Power[a, -1]`?

Reply:
I don't have access to Mathematica.

Reply:
Me neither